File: fetchPosts.js
const fs = require('fs').promises;
const path = require('path');

// Load the markdown-it library from the local file
const MarkdownIt = require('./libs/markdown-it'); // Adjust the path as necessary
const md = new MarkdownIt(); // Create an instance of markdown-it

const postsDirPath = path.join(__dirname, 'posts'); // Directory containing JSON and MD files
const MAX_CONCURRENT_WRITES = 200; // Maximum concurrent writes

async function fetchPosts() {
    try {
        const allPosts = [];
        let totalPages = 0; // Total pages count
        let nonPageFiles = 0; // Count of non-page files
        let staticFiles = 0; // Count of static files

        // Function to read JSON and MD files recursively
        async function readFiles(dir) {
            const files = await fs.readdir(dir, { withFileTypes: true });
            for (const file of files) {
                const filePath = path.join(dir, file.name);
                if (file.isDirectory()) {
                    // Recursively read subdirectories
                    await readFiles(filePath);
                } else if (file.isFile()) {
                    if (file.name.endsWith('.json')) {
                        // Read and copy JSON files as raw files
                        const data = await fs.readFile(filePath, 'utf8');
                        const folder = path.relative(postsDirPath, dir); // Get the folder name
                        allPosts.push({ content: data, folder, isJson: true }); // Mark as JSON
                        totalPages++; // Increment total pages count for JSON files
                        nonPageFiles++; // Increment non-page files count
                    } else if (file.name.endsWith('.md')) {
                        // Read Markdown files
                        const data = await fs.readFile(filePath, 'utf8');
                        const folder = path.relative(postsDirPath, dir); // Get the folder name
                        const content = md.render(data); // Convert Markdown to HTML using markdown-it
                        allPosts.push({ content, folder }); // Add post info without title
                        totalPages++; // Increment total pages count for Markdown files
                    } else {
                        staticFiles++; // Increment static files count for other file types
                    }
                }
            }
        }

        // Start reading from the posts directory
        await readFiles(postsDirPath);

        const outputDir = path.join(__dirname, 'public'); // Directly point to the public directory
        await fs.mkdir(outputDir, { recursive: true });
// Function to format JSON content into HTML
const formatJson = (jsonString) => {
    try {
        const jsonData = JSON.parse(jsonString);
        return `<pre>${JSON.stringify(jsonData, null, 2)}</pre>`; // Format JSON with indentation
    } catch (error) {
        console.error('Error parsing JSON:', error);
        return '<p>Error parsing JSON data.</p>';
    }
};

// Function to write a single post to a file
const writePost = async (post, index) => {
    const { content, folder, isJson } = post;
    const fileName = `post-${index + 1}.html`; // Use incremental numbers for file names

    const folderPath = path.join(outputDir, folder); // Create a path for the folder
    await fs.mkdir(folderPath, { recursive: true }); // Ensure the folder exists

    const filePath = path.join(folderPath, fileName); // Full path for the file

    // Write HTML content for both JSON and Markdown files
    const htmlContent = `<!DOCTYPE html>
<html lang="en">
<head>
    <meta charset="UTF-8">
    <meta name="viewport" content="width=device-width, initial-scale=1.0">
    <title>Post ${index + 1}</title>
</head>
<body>
    <h1>Post ${index + 1}</h1>
    <div>${isJson ? formatJson(content) : content}</div> 
</body>
</html>
`;

    await fs.writeFile(filePath, htmlContent);
};

// The rest of your fetchPosts function remains unchanged

        // Function to write a single post to a file
  // Function to write a single post to a file
const writePost = async (post, index) => {
    const { content, folder, isJson } = post;
    const fileName = `post-${index + 1}.html`; // Use incremental numbers for file names

    const folderPath = path.join(outputDir, folder); // Create a path for the folder
    await fs.mkdir(folderPath, { recursive: true }); // Ensure the folder exists

    const filePath = path.join(folderPath, fileName); // Full path for the file

    // Write HTML content for both JSON and Markdown files
    const htmlContent = `<!DOCTYPE html>
<html lang="en">
<head>
    <meta charset="UTF-8">
    <meta name="viewport" content="width=device-width, initial-scale=1.0">
    <title>Post ${index + 1}</title>
</head>
<body>
    <h1>Post ${index + 1}</h1>
    <div>${isJson ? formatJson(content) : content}</div> 
</body>
</html>
`;

// Function to format JSON content into HTML
const formatJson = (jsonString) => {
    try {
        const jsonData = JSON.parse(jsonString);
        return `<pre>${JSON.stringify(jsonData, null, 2)}</pre>`; // Format JSON with indentation
    } catch (error) {
        console.error('Error parsing JSON:', error);
        return '<p>Error parsing JSON data.</p>';
    }
};

await fs.writeFile(filePath, htmlContent);
};


        // Process posts with limited concurrency
        const processPosts = async () => {
            for (let i = 0; i < allPosts.length; i += MAX_CONCURRENT_WRITES) {
                const chunk = allPosts.slice(i, i + MAX_CONCURRENT_WRITES);
                await Promise.all(chunk.map((post, index) => writePost(post, i + index)));
            }
        };

        // Start processing posts
        const startTime = Date.now();
        await processPosts();
        const endTime = Date.now();

        // Log the statistics
        console.log('--- Build Statistics ---');
        console.log(`Total Posts Created: ${allPosts.length}`);
        console.log(`Total Build Time: ${endTime - startTime} ms`); // Log total build time
               console.log(`Total JSON Files Processed: ${nonPageFiles}`);
        console.log(`Total Static Files Ignored: ${staticFiles}`);

    } catch (err) {
        console.error('Error:', err);
    }
}

// Start the fetchPosts function
fetchPosts();

File: fetchPosts.go
package main

import (
    "crypto/sha256"
    "encoding/hex"
    "encoding/json"
    "fmt"
    "html/template"
    "io/ioutil"
    "os"
    "path/filepath"
    "strings"
    "sync"
    "time"
)

type Post struct {
    Title   string `json:"title"`
    Content string `json:"content"`
    Date    string `json:"date"`
    Folder  string `json:"folder"`
}

var htmlTemplate = template.Must(template.New("post").Parse(`
<!DOCTYPE html>
<html lang="en">
<head>
    <meta charset="UTF-8">
    <meta name="viewport" content="width=device-width, initial-scale=1.0">
    <title>{{.Title}}</title>
</head>
<body>
    <h1>{{.Title}}</h1>
    <div>{{.Content}}</div>
</body>
</html>
`))

func main() {
    startTime := time.Now()
    postsDirPath := "./posts"
    outputDir := "./public"

    os.MkdirAll(outputDir, os.ModePerm)

    var paths []string
    err := filepath.Walk(postsDirPath, func(path string, info os.FileInfo, err error) error {
        if err != nil || info.IsDir() {
            return err
        }
        paths = append(paths, path)
        return nil
    })

    if err != nil {
        fmt.Println("Error walking directory:", err)
        return
    }

    processFiles(paths, outputDir)

    fmt.Printf("Total Build Time: %v\n", time.Since(startTime))
}

func processFiles(paths []string, outputDir string) {
    var wg sync.WaitGroup
    sem := make(chan struct{}, 10) // Limit concurrency to 10 goroutines

    for _, path := range paths {
        sem <- struct{}{} // Acquire semaphore
        wg.Add(1)
        go func(path string) {
            defer wg.Done()
            defer func() { <-sem }() // Release semaphore

            processFile(path, outputDir)
        }(path)
    }

    wg.Wait()
}

func processFile(path string, outputDir string) {
    data, err := ioutil.ReadFile(path)
    if err != nil {
        fmt.Println("Error reading file:", err)
        return
    }

    folder := filepath.Dir(path)
    ext := filepath.Ext(path)
    fileName := strings.TrimSuffix(filepath.Base(path), ext)

    var post Post
    switch ext {
    case ".json":
        var posts []Post
        if err := json.Unmarshal(data, &posts); err != nil {
            fmt.Println("Error parsing JSON:", err)
            return
        }
        for _, p := range posts {
            p.Folder = folder
            generateHTML(p, outputDir)
        }
    case ".md":
        post.Title = fileName
        post.Content = string(data)
        post.Date = time.Now().Format(time.RFC3339)
        post.Folder = folder
        generateHTML(post, outputDir)
    }
}

func generateHTML(post Post, outputDir string) {
    uniqueFileName := generateUniqueFilename(post.Title)
    folderPath := filepath.Join(outputDir, post.Folder)
    os.MkdirAll(folderPath, os.ModePerm)

    filePath := filepath.Join(folderPath, uniqueFileName)
    file, err := os.Create(filePath)
    if err != nil {
        fmt.Println("Error creating file:", err)
        return
    }
    defer file.Close()

    if err := htmlTemplate.Execute(file, post); err != nil {
        fmt.Println("Error executing template:", err)
        return
    }

    fmt.Println("Created post:", filepath.Join(post.Folder, uniqueFileName))
}

func generateUniqueFilename(title string) string {
    hash := sha256.Sum256([]byte(title))
    return hex.EncodeToString(hash[:]) + ".html"
}
